### PR TITLE
[msgs::ChannelReestablish] implement encoding and decoding

### PR DESCRIPTION
**Pull request by savil**
Summary:

Implementing these, taking inspiration from the surrounding code since i'm a rust n00b.

Test Plan:
```
cargo build
cargo test channel_reestablish
```
output:
```
running 3 tests
test ln::msgs::tests::encoding_channel_reestablish_with_secret ... ok
test ln::msgs::tests::decode_channel_reestablish_bad_length ... ok
test ln::msgs::tests::encoding_channel_reestablish_no_secret ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 14 filtered out
```

### CODE REVIEW

**Comment by yuntai** [comment]
hi @savil 
You may want to check `fuzz/fuzz_targets` for msg testing.

**Comment by savil** [comment]
@yuntai ah, thanks!

- I'm looking at `fuzz/fuzz_targets/msg_targets`. I don't 100% follow all of it, but it seems like I would need to add `ChannelReestablish` to `gen_targets.sh`. Is that correct? Is there anything else to add?
- there seems to be some problem with `gen_targets.sh`. If I run it on `master` branch, I get this output: https://gist.github.com/savil/2b3114eaf34c3ef8499b65005e374841. I couldn't spot the problem yet. Am going to sleep but can revisit it over the weekend.

File: src/ln/msgs.rs [hunk 1]
```diff
@@ -1534,3 +1582,85 @@ impl MsgEncodable for OnionErrorPacket {
...
+		};
+
+		let encoded_value = cr.encode();
+		assert_eq!(
+			encoded_value, 
+			vec![4, 0, 0, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 4, 3, 27, 132, 197, 86, 123, 18, 100, 64, 153, 93, 62, 213, 170, 186, 5, 101, 215, 30, 24, 52, 96, 72, 25, 255, 156, 23, 245, 233, 213, 221, 7, 143]
+		);
+
+		// sanity-check that we get the original ChannelReestablish back
+		assert_eq!(msgs::ChannelReestablish::decode(&encoded_value), Ok(cr));
```
Comment (TheBlueMatt) [inline src/ln/msgs.rs]: This should be pretty well covered by the fuzz tests.

**Comment by savil** [comment]
ack. Made requested fixes:
- added fuzz testing 
- focussed new unit tests I added to verify outputs with standard encodings

File: src/ln/msgs.rs [hunk 2]
```diff
@@ -1075,13 +1076,60 @@ impl MsgEncodable for UpdateFee {
 }
 
 impl MsgDecodable for ChannelReestablish {
-	fn decode(_v: &[u8]) -> Result<Self, DecodeError> {
-		unimplemented!();
+	fn decode(v: &[u8]) -> Result<Self, DecodeError> {
+		// TODO savil. Remove comment, after PR discussion.
+		// Why is PublicKey len == 33, shouldn't it be 64? https://git.io/fNWqR
```
Comment (savil) [inline src/ln/msgs.rs]: care to shed light on this question?

Comment (TheBlueMatt) [reply src/ln/msgs.rs]: No, all pubkeys are encoded in "compressed form", see https://docs.rs/secp256k1/0.9.2/secp256k1/key/struct.PublicKey.html#method.serialize

File: src/ln/msgs.rs [hunk 3]
```diff
@@ -1534,3 +1582,57 @@ impl MsgEncodable for OnionErrorPacket {
...
+			next_local_commitment_number: 3,
+			next_remote_commitment_number: 4,
+			your_last_per_commitment_secret: Some([9; 32]),
+			my_current_per_commitment_point: public_key,
+		};
+
+		let encoded_value = cr.encode();
+		assert_eq!(
+			encoded_value, 
+			vec![4, 0, 0, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 4, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 3, 27, 132, 197, 86, 123, 18, 100, 64, 153, 93, 62, 213, 170, 186, 5, 101, 215, 30, 24, 52, 96, 72, 25, 255, 156, 23, 245, 233, 213, 221, 7, 143],
```
Comment (TheBlueMatt) [inline src/ln/msgs.rs]: Weird, rust 1.22.0 barfed here, not really sure why.

**Comment by TheBlueMatt** [comment]
Also you need to add the new bin "msg_channel_reestablish_target" to fuzz/Cargo.toml.

**Comment by TheBlueMatt** [comment]
Will take as #75 once travis runs there.

**Pull request closed without merging**